Bring README up to date with `master` branch (#247)
awslabs/aws-lambda-rust-runtime: README.md

This package makes it easy to run AWS Lambda Functions written in Rust. This workspace includes multiple crates:

> - [![Docs](https://docs.rs/lambda/badge.svg)](https://docs.rs/lambda) **`lambda`** is a library that makes it easy to write Lambda functions in Rust.

Reviewer: Should we say somewhere that this is actually the runtime that runs the handler function?

It may be good say that the runtime and the handler function compile into a single executable called bootstrap that is loaded by Lambda and then runs in a loop waiting for an event.

All this may be obvious to those working on the runtime, but an outsider has no idea how it all works inside AWS. Other runtimes like Node or Java only deal with handler functions, so it's likely to be a step up in complexity for anyone landing here for the first time.

Reviewer 2: Suggestion that might add clarity

```
... provides a Lambda runtime for applications written in Rust.
```

Author: This is a great callout, and also addresses that this line just repeats what is said in the header above. It makes the fact that this is a provided runtime more explicit, which might also help people when they get to deployment. Will update.

Author: Likely need to update the phrasing at line 146 as well

Right now I have:

```suggestion
- [![Docs](https://docs.rs/lambda/badge.svg)](https://docs.rs/lambda) **`lambda`** is a library that provides a Lambda runtime for applications written in Rust.
```
from @softprops and 

> With the `lambda` library you can build your Rust application to implement an [AWS Lambda's custom runtime](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html)

which I was considering at Line 146.

- [![Docs](https://docs.rs/lambda_http/badge.svg)](https://docs.rs/lambda_http) **`lambda-http`** is a library that makes it easy to write API Gateway proxy event focused Lambda functions in Rust.

> - [![Docs](https://docs.rs/lambda_attributes/badge.svg)](https://docs.rs/lambda_attributes) **`lambda-attributes`** holds an [attribute macro](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros) that runs an `async main` function in the Lamda runtime. You probably don't need to use this crate directly, the macro is exposed by **`lambda`**!

Reviewer 2: Listing this might add more cognitive overhead than is useful for users. I don't think this crate is designed to be depended on directly.

Author: I left it for parity with the previous README that listed each member crate of the workspace. Completely open to removing it, I agree it seems like an implementation detail that is best addressed in the docs rather than the README.

Reviewer: Well, if there is a crate I'd be asking myself, do I need it or what?
Shall we clearly state that:
```
The macro is exposed by **`lambda`**. You do not need to use this crate directly.
```

## Example function

> The code below creates a simple function that receives an event and echoes it back as a response. Notice: to run these examples, we require a minimum Rust version of 1.31.

Reviewer 3: An MSRV of 1.31 is _not_ correct. Right now, it's _at least_ a 1.42_, but I'd for the time being, I'd like to make the policy "latest stable Rust" until this reaches 1.0.

Reviewer 2: + 1 on stable rust as a contract

Reviewer 2: \+ 1 (because github markdown)

Author: How does:

> Notice: this crate is tested against latest stable Rust.

sound?

```rust,no_run
use lambda::{lambda, Context};
use serde_json::Value;

type Error = Box<dyn std::error::Error + Send + Sync + 'static>;

#[lambda]
#[tokio::main]
async fn main(event: Value, _: Context) -> Result<Value, Error> {
    Ok(event)

> }

Reviewer 3: I've been thinking about this, but I think we should cut the `#[lambda]` macro from the initial release. It's problematic when used with anything that expects to be initialized a single time (e.g., a logger) but a _warm_ invoke occurs. In the case of a warm invoke, this will cause the runtime to panic.

Let's prefer a more full example?

Author: Interesting, I did not know this about the attribute macro. My primary reason for avoiding it has been more do with developer tooling, I found `rust-analyser` and `cargo build` output regarding errors to be much less useful when using it.

That type of error (panic on warm invocation) does sound like a reason to hold it back from the release. I can update the README to reflect that.

Reviewer 3: > Interesting, I did not know this about the attribute macro. My primary reason for avoiding it has been more do with developer tooling, I found rust-analyser and cargo build output regarding errors to be much less useful when using it.

That's another compelling reason to not release and I completely forgot about it. I think the `#[lambda]` macro is one of those things that looks nice on a slide/screenshot, but is something most customers will quickly outgrow.

Reviewer 2: >  I did not know this about the attribute macro.

this might just be a matter of documentation. I mentioned this in the lambda http docs as a tradeoff to consider. I still think its useful for specific cases but it's also good to be mindful of the tradeoffs.

```

> The code above is the same as the [basic example](https://github.com/awslabs/aws-lambda-rust-runtime/tree/master/lambda-runtime/examples/hello.rs) in the `lambda` crate. An [alternative example](https://github.com/awslabs/aws-lambda-rust-runtime/tree/master/lambda-runtime/examples/hello-without-macros.rs) that uses functions instead of an attribute macro to create the handler is available as well.

Reviewer: The links to examples may change after PR #244 is completed.

### Deployment

> There are currently multiple ways of building this package: manually with the AWS CLI, and with the [Serverless framework](https://serverless.com/framework/).

Reviewer 3: Let's maybe include the SAM CLI with [the Makefile approach](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-custom-runtimes.html) and (hopefully) https://github.com/awslabs/aws-lambda-builders/pull/174, once that is merged.

Reviewer 2: We'll see... :) I'll try to keep a some light pressure on that. Ultimately SAM is the most likely AWS-suggested approach to expect with SAM being the tool AWS pushes. Not having a SAM story is worse than not having some SAM story. I agree with the adding SAM + Makefile example being useful

![s551228258327922884_p1702_i14_w300](https://user-images.githubusercontent.com/2242/88725720-829dc300-d0fa-11ea-8110-8f4a19fbd548.jpg)

Reviewer 4: I'd much more believe in terraform ;d, could add an example with that

Reviewer: > I'd much more believe in terraform ;d, could add an example with that
I prefer TF too, but if AWS prefers SAM then what most people will use anyway.
@Veetaha if you do make an example, can we put it in examples folder with a mention and a link to it from README to keep it concise?

Reviewer 5: Perhaps putting a dissonant note on this thread, I'm using CDK for deployment of Rust lambdas these days XD

If people adopt SAM as a Rust-lambda "blueprint project", there I'll go too... My current DX fail with Rust lambdas is that I've been unable to find a lightweight and [properly documented way to debug Rust lambdas locally](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-debugging.html)... 

My ideal scenario: not having to deal with docker (i.e: lambci) and my VSCode session having **all the variable values, watchpoints and backtraces I would normally have for "regular local binaries"**, ideally with proper observability all the way down to FFI (bindgen) crates.

Am I dreaming or is this available today and I've missed it? :)

Reviewer 5: Essentially I would like something like this for Rust+SAM+lambdas: https://pawelgrzybek.com/attach-visual-studio-code-debugger-to-sam-serverless-application-model-local-endpoint/ ... does anybody has such a setup? @rimutaka ?

Reviewer: @brainstorm,  That debug-proxy works for me just fine and I use CLI/TF for deployment.

```bash
$ aws lambda invoke --function-name rustTest \
  --payload '{"Hello,": "world!"}' \
  output.json

> $ cat output.json  # Prints: {"Hello,": "world!"}

Reviewer 3: I'm not sure I'm in favor of this chance. Can we prefer the original?

Author: I've added the previous functionality back to the README and the `hello-without-macro` example that it references. It is working on and generating an untyped `json` value. Let me know if you'd like to add custom input and output events back to this headline example as well. Leaving the discussion on custom events to later in the documentation allows a gentler introduction in my opinion.

# Ctrl-D to yield control back to your function
```

## `lambda`

> This library makes it easy to create Rust executables for AWS lambda. The library defines an `#[lambda]` attribute macro. Adding the `#[lambda]` attribute to your `main` function allows you to define your event handler logic in one function as shown in the example above.

Reviewer 3: ```suggestion
`lambda` is a library for authoring reliable and performant Rust-based AWS Lambda functions. At a high level, it provides a few major components:

- `Handler`, a trait that defines interactions between customer-authored code and this library.
- `lambda::run`, function that runs an `Handler`.
```

Reviewer: A lay user perspective: can the Handler part be described in terms consistent with other runtimes?

- `Handler`, a trait that defines interactions between customer-authored code (the handler function described in other runtimes) and this library (the runtime).

- `lambda::run`, function that runs an `Handler` and interacts with the rest of the AWS infrastructure to fetch new event data and return the response.

> This library makes it easy to create Rust executables for AWS lambda. The library defines exposes an `#[lambda]` attribute macro. Adding the `#[lambda]` attribute to your `main` function allows you to define your event handler logic in one function as show in the example above.

Reviewer: as show -> as shown

Author: Thanks - will address. It's funny how many times you can read something and miss things like this.

> It also exposes the `Handler` trait. A type that conforms to this trait can be passed to the `lambda::run` function, which launches and runs the Lambda runtime.

Reviewer 3: Let's remove this.

> The helper function `handler_fn` provides a default implementation that enables you to provide a Rust closure or function pointer to the `lambda::run` function.

Reviewer 3: Two notes:

- Let's avoid the second person tense.
- Provide an example of using this?

```suggestion
The function `handler_fn` converts a rust function or closure to `Handler`, which can then be run by `lambda::run`.
```

```rust
extern crate serde;

> extern crate serde_json;

Reviewer 4: I recommend removing `extern crate` entries too
